etl/files.py

```diff
@@ -4,23 +4,37 @@
 
 import hashlib
 from pathlib import Path
-from typing import List, Set, Union
+from threading import Lock
+from typing import Dict, List, Set, Union
+
+# runtime cache, we need locks because we usually run it in threads
+cache_md5: Dict[str, str] = {}
+cache_md5_locks: Dict[str, Lock] = {}
 
 
 def checksum_file(filename: Union[str, Path]) -> str:
     "Return the md5 hex digest of the file contents."
     if isinstance(filename, Path):
         filename = filename.as_posix()
 
-    chunk_size = 2**20
-    _hash = hashlib.md5()
-    with open(filename, "rb") as istream:
-        chunk = istream.read(chunk_size)
-        while chunk:
-            _hash.update(chunk)
-            chunk = istream.read(chunk_size)
+    # add checksum to cache
+    if filename not in cache_md5:
+        cache_md5_locks[filename] = Lock()
+
+        with cache_md5_locks[filename]:
+            chunk_size = 2**20
+            _hash = hashlib.md5()
+            with open(filename, "rb") as istream:
+                chunk = istream.read(chunk_size)
+                while chunk:
+                    _hash.update(chunk)
+                    chunk = istream.read(chunk_size)
+
+            md5 = _hash.hexdigest()
```

Comment:
Move this into a single higher level method `checksum_file_nocache()` that you can call here.